### DISCUSSION TITLE
HOW TO CONNECT TO GCS FROM KEYCLOAK DOCKER CONTAINER?

### DISCUSSION THREAD

**Discussion by hiddenmyth** [keycloak/keycloak#25366] [Q&A]
### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

storage

### Describe the bug

Hi
I am using keycloak container quay.io/keycloak/keycloak:latest on google kubernetes engine
and trying to get access into google cloud sql with public ip address

here comes my environment variables

   -----------------------------------------
    - name: KC_DB
      value: postgres
    - name: KC_DB_URL_HOST
      value: <GCS_INSTANCE_PUBLIC_IP_V4_ADDRESS>
    - name: KC_DB_URL_PORT
      value: "5432"
    - name: KC_DB_URL_DATABASE
      value: keycloak
    - name: KC_DB_SCHEMA
      value: public
... ... ...
   -----------------------------------------

But it fails, with this error when I start the container from Google kubernetes engine cluster

WARN [io.agroal.pool] (agroal-11) Datasource '': The connection attempt failed."

ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection"

How can I use cloud sql with current stacked keycloak?

### Version

latest

### Expected behavior

NEED TO CONNECT TO GOOGLE CLOUD SQL POSTGRES INSTANCE

### Actual behavior

> WARN [io.agroal.pool] (agroal-11) Datasource '': The connection attempt failed."
> 
> ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection"

### How to Reproduce?

apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: keyclaok
  name: keycloak
  namespace: default
spec:
  podManagementPolicy: Parallel
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: keycloak
  serviceName: keycloak-headless
  template:
    metadata:
      labels:
        app: keycloak
    spec:
      containers:
      - env:
        - name: KC_HOSTNAME_URL
          valueFrom:
            secretKeyRef:
              key: KC_HOSTNAME_URL
              name: keycloak-secrets
        - name: KC_HOSTNAME_ADMIN_URL
          valueFrom:
            secretKeyRef:
              key: KC_HOSTNAME_ADMIN_URL
              name: keycloak-secrets
        - name: KEYCLOAK_ADMIN
          valueFrom:
            secretKeyRef:
              key: KEYCLOAK_ADMIN
              name: keycloak-secrets
        - name: KEYCLOAK_ADMIN_PASSWORD
          valueFrom:
            secretKeyRef:
              key: KEYCLOAK_ADMIN_PASSWORD
              name: keycloak-secrets
        - name: KC_DB
          value: postgres
        - name: KC_DB_URL_HOST
          value: <PUBLIC IP ADDRESS>
        - name: KC_DB_URL_PORT
          value: "5432"
        - name: KC_DB_URL_DATABASE
          value: keycloak
        - name: KC_DB_SCHEMA
          value: public
        - name: KC_DB_USERNAME
          valueFrom:
            secretKeyRef:
              key: POSTGRES_USER
              name: postgres-sec
        - name: KC_DB_PASSWORD
          valueFrom:
            secretKeyRef:
              key: POSTGRES_PASSWORD
              name: postgres-sec
        - name: KC_HOSTNAME_STRICT_HTTPS
          value: "false"
        - name: KC_HTTP_ENABLED
          value: "true"
        - name: KC_PROXY
          value: "edge"
        image: quay.io/keycloak/keycloak:latest
        args: ["start"]
        imagePullPolicy: IfNotPresent
        name: keycloak
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        - containerPort: 8443
          name: https
          protocol: TCP
        resources: {}
        securityContext:
          runAsNonRoot: true
          runAsUser: 1000
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      terminationGracePeriodSeconds: 60
  updateStrategy:
    type: RollingUpdate

**Comment by robson90** [comment]
Hey, the problem could be many things. 

Do you have a Service Account, that is able to Talk to the Google API´s?
In my Project I had to use Workload Identity. Read here: https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity?hl=de

We did our setup with Terraform.
resource "kubernetes_service_account" -> have service account available inside the Cluster
resource "google_service_account" -> Creating the service account
resource "google_project_iam_member" -> give the service account permissions, in our case we gave it "roles/storage.admin"
resource "google_service_account_iam_member" -> give the service account of the Cluster, the permission to impersonate as the google service account.

Hope that helps and points you into the right direction. If not, feel free to ask.


Looking for example configuration of Keycloak in Kubernetes ? Habe a look here: https://github.com/bitnami/charts/tree/main/bitnami/keycloak

**Reply by hiddenmyth** [reply]
Hi @robson90 
Ty for your response
Anyway, I already use it and exposed as public address
Then I can connect via Navicat
But can't connect from Keycloak

**Reply by hiddenmyth** [reply]
As far as I opened my CloudSQL instance with public IP address with authorized ip list then I am able to connect from my navicat, then why keycloak is not able to connect through?
And to tell about the service account, yes I have and already set permission as full

**Reply by robson90** [reply]
Im sorry, for that. I'm not that good at the cloud stuff. 
Just for reference.(MySQL, Postgres)
https://cloud.google.com/sql/docs/mysql/connect-kubernetes-engine

Wish you luck!
